### PR TITLE
Events in federated namespace controller

### PR DESCRIPTION
**Pull request by mwielgus**
@quinton-hoole @nikhiljindal @wojtek-t @kubernetes/sig-cluster-federation

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31640)

### CODE REVIEW

File: federation/pkg/federation-controller/namespace/namespace_controller.go [hunk 1]
```diff
@@ -290,6 +308,9 @@ func (nc *NamespaceController) reconcileNamespace(namespace string) {
 	}
 	err = nc.federatedUpdater.Update(operations, nc.updateTimeout)
 	if err != nil {
+		nc.eventRecorder.Event(baseNamespace, api.EventTypeWarning, "FailedUpdate",
+			"Failed to execute updates")
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: We should say which update failed and why

File: federation/pkg/federation-controller/namespace/namespace_controller.go [hunk 2]
```diff
@@ -290,6 +308,9 @@ func (nc *NamespaceController) reconcileNamespace(namespace string) {
 	}
 	err = nc.federatedUpdater.Update(operations, nc.updateTimeout)
 	if err != nil {
+		nc.eventRecorder.Event(baseNamespace, api.EventTypeWarning, "FailedUpdate",
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: FailedClusterUpdate?

File: federation/pkg/federation-controller/namespace/namespace_controller.go [hunk 3]
```diff
@@ -264,6 +276,9 @@ func (nc *NamespaceController) reconcileNamespace(namespace string) {
 		}
 
 		if !found {
+			nc.eventRecorder.Eventf(baseNamespace, api.EventTypeNormal, "UpdateCluster",
+				"Create namespace in cluster %s", cluster.Name)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Its better to be explicit about is this Creat*ed* or Creat*ing*. I think it should be Creating here.

File: federation/pkg/federation-controller/namespace/namespace_controller.go [hunk 4]
```diff
@@ -74,13 +80,19 @@ type NamespaceController struct {
 
 // NewNamespaceController returns a new namespace controller
 func NewNamespaceController(client federation_release_1_4.Interface) *NamespaceController {
+
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(eventsink.NewFederatedEventSink(client))
+	recorder := broadcaster.NewRecorder(api.EventSource{Component: "namespace-controller"})
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Not sure where this is used and should this be unique across all controllers, but Component name "federation-namespace-controller" should be better?

**Comment by nikhiljindal** [comment]
Not sure whats the rule you are using around when to generate an event. Should we generate an event each time there is an error?
Also should we generate an event when a namespace is deleted? (We are generating an event when it is marked terminating)

File: federation/pkg/federation-controller/namespace/namespace_controller.go [hunk 5]
```diff
@@ -74,13 +80,19 @@ type NamespaceController struct {
 
 // NewNamespaceController returns a new namespace controller
 func NewNamespaceController(client federation_release_1_4.Interface) *NamespaceController {
+
```
Comment (wojtek-t) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: nit: please remove empty line

File: federation/pkg/federation-controller/namespace/namespace_controller.go (hunk 4 above)
Comment (wojtek-t) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: +1 to include something about federation

File: federation/pkg/federation-controller/namespace/namespace_controller.go [hunk 6]
```diff
@@ -264,6 +276,9 @@ func (nc *NamespaceController) reconcileNamespace(namespace string) {
 		}
 
 		if !found {
+			nc.eventRecorder.Eventf(baseNamespace, api.EventTypeNormal, "UpdateCluster",
```
Comment (wojtek-t) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Actually, why is "reason" set to "UpdateCluster"?

File: federation/pkg/federation-controller/namespace/namespace_controller_test.go [hunk 7]
```diff
@@ -75,7 +75,8 @@ func TestNamespaceController(t *testing.T) {
 
 	ns1 := api_v1.Namespace{
 		ObjectMeta: api_v1.ObjectMeta{
-			Name: "test-namespace",
+			Name:     "test-namespace",
+			SelfLink: "/api/v1/namespaces/test-namespace",
```
Comment (wojtek-t) [inline federation/pkg/federation-controller/namespace/namespace_controller_test.go]: Hmm - why do you need this?

File: federation/pkg/federation-controller/namespace/namespace_controller.go [hunk 8]
```diff
@@ -309,6 +330,7 @@ func (nc *NamespaceController) delete(namespace *api_v1.Namespace) {
 		},
 	}
 	if namespace.Status.Phase != api_v1.NamespaceTerminating {
+		nc.eventRecorder.Event(namespace, api.EventTypeNormal, "Delete", fmt.Sprintf("Marking for deletion"))
```
Comment (wojtek-t) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Delete<Something> (Delete itself is not very informative...)

File: federation/pkg/federation-controller/namespace/namespace_controller.go (hunk 1 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Done. Needed #31641 to do this.

File: federation/pkg/federation-controller/namespace/namespace_controller.go (hunk 2 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Done.

File: federation/pkg/federation-controller/namespace/namespace_controller.go (hunk 3 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Done.

File: federation/pkg/federation-controller/namespace/namespace_controller.go (hunk 4 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Done.

File: federation/pkg/federation-controller/namespace/namespace_controller.go (hunk 5 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Done.

File: federation/pkg/federation-controller/namespace/namespace_controller.go (hunk 6 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Done.

File: federation/pkg/federation-controller/namespace/namespace_controller_test.go (hunk 7 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/namespace/namespace_controller_test.go]: Events need selflink for proper attaching.

File: federation/pkg/federation-controller/namespace/namespace_controller.go (hunk 8 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Done.

**Comment by mwielgus** [comment]
@nikhiljindal We should generate an event when there is an explicit problem with a namespace, not with controller, syncing, etc.

**Comment by wojtek-t** [comment]
LGTM - but please fix tests :-)

**Comment by mwielgus** [comment]
Yeah, I the ran tests in a wrong directory :-/. Fixed.

**Pull request merged**